lib/Dialect/FSM/FSMOps.cpp

```diff
@@ -127,6 +150,13 @@ LogicalResult MachineOp::verify() {
   return success();
 }
 
+LogicalResult MachineOp::canonicalize(MachineOp op, PatternRewriter &rewriter) {
+  // Remove any unreachable states.
+  for (auto unreachable : unreachableStates(op))
+    rewriter.eraseOp(unreachable);
+  return success();
+}
```

Comment:
I've started to see some intermittent failures that I believe is coming from this. Looking at this, I'm not sure it is safe to erase _other_ ops from a canonicalizer. In order to do this, I think this would need to be in a separate pass.

I can trip the intermittent failure with:
```sh
while circt-opt --canonicalize ../test/Dialect/FSM/canonicalize.mlir; do echo "not yet"; done
```

The failure looks like:
```
Stack dump:
0.	Program arguments: /Users/schuylere/repos/github.com/llvm/circt/build/bin/circt-opt --canonicalize /Users/schuylere/repos/github.com/llvm/circt/test/Dialect/FSM/canonicalize.mlir
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  circt-opt                0x000000010c5ab31d llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 61
1  circt-opt                0x000000010c5ab85b PrintStackTraceSignalHandler(void*) + 27
2  circt-opt                0x000000010c5a97b6 llvm::sys::RunSignalHandlers() + 134
3  circt-opt                0x000000010c5ad1ef SignalHandler(int) + 223
4  libsystem_platform.dylib 0x00007ff81bd85dfd _sigtramp + 29
5  circt-opt                0x000000010c34d599 mlir::TypeID::getFromOpaquePointer(void const*) + 25
6  circt-opt                0x000000010ce0d0f2 circt::fsm::StateOp::transitions() + 34
7  circt-opt                0x000000010ce0cf22 circt::fsm::StateOp::getNextStates() + 82
8  circt-opt                0x000000010ce0bdff unreachableStates(circt::fsm::MachineOp) + 239
9  circt-opt                0x000000010ce0bbf6 circt::fsm::MachineOp::canonicalize(circt::fsm::MachineOp, mlir::PatternRewriter&) + 54
10 circt-opt                0x000000010ce3166e mlir::RewritePatternSet& mlir::RewritePatternSet::add<circt::fsm::MachineOp>(mlir::LogicalResult (*)(circt::fsm::MachineOp, mlir::PatternRewriter&))::FnPattern::matchAndRewrite(circt::fsm::MachineOp, mlir::PatternRewriter&) const + 46
11 circt-opt                0x000000010ce315cb mlir::detail::OpOrInterfaceRewritePatternBase<circt::fsm::MachineOp>::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&) const + 59
12 circt-opt                0x000000010e950146 mlir::PatternApplicator::matchAndRewrite(mlir::Operation*, mlir::PatternRewriter&, llvm::function_ref<bool (mlir::Pattern const&)>, llvm::function_ref<void (mlir::Pattern const&)>, llvm::function_ref<mlir::LogicalResult (mlir::Pattern const&)>) + 1750
13 circt-opt                0x000000010e8dc6a9 (anonymous namespace)::GreedyPatternRewriteDriver::simplify(llvm::MutableArrayRef<mlir::Region>) + 2137
14 circt-opt                0x000000010e8dbcef mlir::applyPatternsAndFoldGreedily(llvm::MutableArrayRef<mlir::Region>, mlir::FrozenRewritePatternSet const&, mlir::GreedyRewriteConfig) + 319
15 circt-opt                0x000000010dcfd315 (anonymous namespace)::Canonicalizer::runOnOperation() + 101
16 circt-opt                0x000000010eaae7eb mlir::detail::OpToOpPassAdaptor::run(mlir::Pass*, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int) + 587
17 circt-opt                0x000000010eaaeed0 mlir::detail::OpToOpPassAdaptor::runPipeline(llvm::iterator_range<llvm::pointee_iterator<std::__1::unique_ptr<mlir::Pass, std::__1::default_delete<mlir::Pass>>*, mlir::Pass>>, mlir::Operation*, mlir::AnalysisManager, bool, unsigned int, mlir::PassInstrumentor*, mlir::PassInstrumentation::PipelineParentInfo const*) + 352
18 circt-opt                0x000000010eab0d56 mlir::PassManager::runPasses(mlir::Operation*, mlir::AnalysisManager) + 134
19 circt-opt                0x000000010eab0b82 mlir::PassManager::run(mlir::Operation*) + 882
20 circt-opt                0x000000010dc306ed performActions(llvm::raw_ostream&, bool, bool, llvm::SourceMgr&, mlir::MLIRContext*, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>) + 445
21 circt-opt                0x000000010dc2ecc3 processBuffer(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, bool, bool, bool, bool, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, mlir::DialectRegistry&, llvm::ThreadPool*) + 467
22 circt-opt                0x000000010dc2ea15 mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, llvm::function_ref<mlir::LogicalResult (mlir::PassManager&)>, mlir::DialectRegistry&, bool, bool, bool, bool, bool) + 485
23 circt-opt                0x000000010dc2ee8a mlir::MlirOptMain(llvm::raw_ostream&, std::__1::unique_ptr<llvm::MemoryBuffer, std::__1::default_delete<llvm::MemoryBuffer>>, mlir::PassPipelineCLParser const&, mlir::DialectRegistry&, bool, bool, bool, bool, bool) + 234
24 circt-opt                0x000000010dc2faf5 mlir::MlirOptMain(int, char**, llvm::StringRef, mlir::DialectRegistry&, bool) + 3093
25 circt-opt                0x000000010c34cc68 main + 200
26 dyld                     0x0000000122bb151e start + 462
```

Reply:
Hmm... not good! I'll try to move this to be a canonicalization of the `StateOp` itself and see whether that fixes the crash.